--- quickjs-libc-orig.c	2022-08-27 09:33:58.000000000 -0700
+++ ./quickjs-libc.c	2022-08-27 09:11:00.000000000 -0700
@@ -65,8 +65,13 @@
 #endif
 #endif /* __APPLE__ */
 
+#if defined(__FreeBSD__)
+typedef sig_t sighandler_t;
+extern char **environ;
 #endif
 
+#endif /* _WIN32 */
+
 #if !defined(_WIN32)
 /* enable the os.Worker API. IT relies on POSIX threads */
 /*#define USE_WORKER*/
@@ -622,7 +627,7 @@
   #define NATIVE_LIBRARY_SUFFIX ".dll"
 #elif defined(__APPLE__)
   #define NATIVE_LIBRARY_SUFFIX ".dylib"
-#elif defined(__linux__)
+#elif defined(__linux__) || defined(__FreeBSD__)
   #define NATIVE_LIBRARY_SUFFIX ".so"
 #endif
 JSModuleDef *js_module_loader(JSContext *ctx,
