--- Makefile-orig	2022-08-27 09:33:58.000000000 -0700
+++ ./Makefile	2022-08-27 09:18:28.000000000 -0700
@@ -25,6 +25,9 @@
 ifeq ($(shell uname -s),Darwin)
 CONFIG_DARWIN=y
 endif
+ifeq ($(shell uname -s),FreeBSD)
+CONFIG_FREEBSD=y
+endif
 # Windows cross compilation from Linux
 #CONFIG_WIN32=y
 # use link time optimization (smaller and faster executables but slower build)
@@ -39,6 +42,9 @@
 CONFIG_CLANG=y
 CONFIG_DEFAULT_AR=y
 endif
+ifdef CONFIG_FREEBSD
+CONFIG_CLANG=y
+endif
 
 # installation directory
 prefix=/usr/local
