--- ../../../../external/re2_2015-11-01/re2.vcxproj	2016-01-28 11:53:18.816326500 -0800
+++ ./re2.vcxproj	2016-01-28 12:16:30.762750400 -0800
@@ -139,7 +139,7 @@
   </ItemDefinitionGroup>
   <ItemGroup>
     <ClCompile Include="re2\*.cc" />
-    <ClCompile Include="util\*.cc" />
+    <ClCompile Include="util\*.cc" Exclude="util\thread.cc;util\test.cc;util\benchmark.cc" />
   </ItemGroup>
   <ItemGroup>
     <ClInclude Include="re2\*.h" />
