diff -ruN --label original original ./include/v8.h
--- original
+++ ./include/v8.h	2017-04-17 21:54:30.000000000 -0700
@@ -854,6 +854,7 @@
 
   // Local::New uses CreateHandle with an Isolate* parameter.
   template<class F> friend class Local;
+  template<class F> friend class Handle;
 
   // Object::GetInternalField and Context::GetEmbedderData use CreateHandle with
   // a HeapObject* in their shortcuts.
