diff --git a/public/common/unicode/unistr.h b/public/common/unicode/unistr.h
index 57ea346..9aba6f0 100644
--- a/public/common/unicode/unistr.h
+++ b/public/common/unicode/unistr.h
@@ -4023,6 +4023,7 @@ UnicodeString::replace(int32_t start,
   int32_t count = 0;
   UBool isError = FALSE;
   U16_APPEND(buffer, count, U16_MAX_LENGTH, srcChar, isError);
+  (void) isError;
   return doReplace(start, _length, buffer, 0, count);
 }
 
@@ -4326,6 +4327,7 @@ UnicodeString::append(UChar32 srcChar) {
   int32_t _length = 0;
   UBool isError = FALSE;
   U16_APPEND(buffer, _length, U16_MAX_LENGTH, srcChar, isError);
+  (void) isError;
   return doReplace(length(), 0, buffer, 0, _length);
 }
 
